### ISSUE TITLE
WebSockets

### ISSUE THREAD

**Issue by KhafraDev** [nodejs/undici#1811]
# `permessage-deflate` support

- Is this something we *need*/*want* to support? Servers that are setup correctly should work even if the client doesn't support it.
- Memory fragmentation issues when using zlib?
- Added complexity of decompressing frames asynchronously

---

# Performance

- Consume the least amount of bytes possible, rather than concatenating every chunk available. See [ws'](https://github.com/websockets/ws/blob/fb1dfd217861757d60c1d02c6e66b4da3630cc93/lib/receiver.js#L94) implementation.
- Switch to Buffer.allocUnsafe in [`lib/websocket/frame.js`](https://github.com/nodejs/undici/blob/860c2617ba55d911286f967dbc915f03e89e0d94/lib/websocket/frame.js#L31)
- Handle TODO comments labeled as "optimize this".
- Benchmarks

---

# Tests

- Add ws' test suite. Note this isn't easy because ws' api does not strictly follow the spec and it occasionally uses internal, underscored properties. Also note that a lot of the validation tests are already handled by the WPTs.
- 100% code coverage for lib/websockets
- Test more strange/error conditions:
    * Chunks that contain thousands of frames
    * Chunks that receive a pong/close frame in the middle of a fragmented message
    * Sending invalid frames
 
---

# Bugs

- using `WebSocket.send` with a Blob asynchronously writes the blob data to the socket. This can cause issues when concurrently sending a blob with anything else. Note: we need support in node core to read a Blob synchronously.
- `ByteParser.run` runs recursively, meaning the max call stack can be exceeded under certain conditions (ie. receiving thousands of frames in a single chunk).

**Comment by jimmywarting** [comment]
> Note: we need support in node core to read a Blob synchronously.

Hmm, would that be really be any good?
if we ever get something like blob backed up by the fs and you would be able to create them from a network mounted cloud drive. then that would block quite a bit.

**Comment by KhafraDev** [comment]
> would that be really be any good?

No, but it's the only way that issue could be fixed without adding complexity.

**Comment by jimmywarting** [comment]
where thinking about trying out websocket today but couldn't find any documentation of how to use it.
eg how to upgrade a request connection

**Comment by KhafraDev** [comment]
Undici only implements a websocket client, the spec one that has documentation on [mdn](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket). If you want to use the lower level dispatcher api from undici, you can use the [`onUpgrade`](https://github.com/nodejs/undici/blob/main/docs/api/Dispatcher.md#parameter-dispatchhandler) callback.

**Comment by jimmywarting** [comment]
Hmm, thanks for that helpful documentation, will try it out, it would be pretty sweet if there where a upgrade method to turn a connection into a WebSocket

**Comment by mcollina** [comment]
Implementing a server is definitely out of scope: use `ws` - none of the spec-compliant work has the settings needed to implement a robust WebSocket server. On another note, we should likely document this somewhere, as this question would come up quite often.

**Comment by jimmywarting** [comment]
> none of the spec-compliant work has the settings or APIs needed to implement a robust WebSocket server

What would it take to maybe say: Build a own npm package that depends on undici WebSocket and tries to build a compatible server? would it be even at all possible by using `import xyz from 'undici/lib/websocket/*'`? and would it require a lot of work?

I'm guessing it would have to work quite differently if you used http/1/2/3 quick WebTransport or whatever is the new "thing"

> we should likely document this somewhere, as this question would come up quite often.

I bet so too. Guess ppl are going to ask about this when they learn that there is a built in WebSocket later

**Comment by yume-chan** [comment]
Is there any updates on the "Setting an undici Dispatcher rather than using the global dispatcher by default." feature?

I need to send each WebSockets to different targets without modifying the `Host` header (the server checks it). By creating multiple dispatchers I can use the `Dispatcher#connect` method to override the target. But now I have to use this hack:

```ts
const agent = new Agent({
    factory(origin, opts) {
        const pool = new Pool(origin, {
            ...opts,
            factory(origin, opts) {
                const client = new Client(origin, opts);
                // Remote debugging validates `Host` header to defend against DNS rebinding attacks.
                // But we can only pass socket name using hostname, so we need to override it.
                (client as any)[Symbols.kHostHeader] = "Host: localhost\r\n";
                return client;
            },
        });
        return pool;
    },
    async connect(options, callback) {
        try {
            const socket = await device.createSocket(
                "localabstract:" + options.hostname
            );
            callback(null, new AdbUndiciSocket(socket) as unknown as Socket);
        } catch (e) {
            callback(e as Error, null);
        }
    },
});
// WebSocket only uses global dispatcher
setGlobalDispatcher(agent);
```

I'm also concerned that undici used in other libraries will also be affected by my global dispatcher.

**Comment by KhafraDev** [comment]
> Is there any updates on the "Setting an undici Dispatcher rather than using the global dispatcher by default." feature?

Yes, I'm planning on implementing my [proposal](https://github.com/whatwg/websockets/issues/42) which will allow us to do so. I should have a PR in the next day, if I remember.
